[java] "Unable to create ValueRepresentation" when using `@LiteralText`

**Affects PMD Version:** *7.19.0*

**Description:**

I configured the XPath rule given below and got the mentioned exception during analysis.

**Exception Stacktrace:**

```
Exception applying rule UseAsManyArgumentsAsPlaceholdersInLogMethod on file [...], continuing with next rule
org.apache.commons.lang3.exception.ContextedRuntimeException: java.lang.RuntimeException: Unable to create ValueRepresentation for value of type: class net.sourceforge.pmd.lang.document.Chars
Exception Context:
	[1:Rule applied on node=!debug only! [MethodCall:113:17]logger.atInfo().log("Got no {} restrictions injected")]
---------------------------------
	at net.sourceforge.pmd.util.AssertionUtil.contexted(AssertionUtil.java:256)
	at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.applyOnIndex(RuleApplicator.java:79)
	at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.apply(RuleApplicator.java:57)
	at net.sourceforge.pmd.lang.rule.internal.RuleSets.apply(RuleSets.java:183)
	at net.sourceforge.pmd.lang.impl.PmdRunnable.processSource(PmdRunnable.java:140)
	at net.sourceforge.pmd.lang.impl.PmdRunnable.run(PmdRunnable.java:80)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
Caused by: java.lang.RuntimeException: Unable to create ValueRepresentation for value of type: class net.sourceforge.pmd.lang.document.Chars
	at net.sourceforge.pmd.lang.rule.xpath.internal.DomainConversion.getAtomicRepresentation(DomainConversion.java:151)
	at net.sourceforge.pmd.lang.rule.xpath.internal.DomainConversion.convert(DomainConversion.java:61)
	at net.sourceforge.pmd.lang.rule.xpath.internal.AstAttributeNode.atomize(AstAttributeNode.java:73)
	at net.sf.saxon.tree.iter.AtomizingIterator.next(AtomizingIterator.java:64)
	at net.sf.saxon.tree.iter.AtomizingIterator.next(AtomizingIterator.java:28)
	at net.sf.saxon.expr.ItemMappingIterator.next(ItemMappingIterator.java:156)
	at net.sf.saxon.expr.CardinalityCheckingIterator.<init>(CardinalityCheckingIterator.java:55)
	at net.sf.saxon.expr.CardinalityChecker.checkCardinality(CardinalityChecker.java:258)
	at net.sf.saxon.expr.CardinalityChecker$CardinalityCheckerElaborator.lambda$elaborateForPull$0(CardinalityChecker.java:506)
	at net.sf.saxon.expr.elab.LazyPullEvaluator.evaluate(LazyPullEvaluator.java:39)
	at net.sf.saxon.expr.FunctionCall$FunctionCallElaborator.evaluateArguments(FunctionCall.java:671)
	at net.sf.saxon.expr.SystemFunctionCall$SystemFunctionCallElaborator.lambda$elaborateForPull$3(SystemFunctionCall.java:628)
	at net.sf.saxon.functions.Count$CountFnElaborator.lambda$elaborateForItem$0(Count.java:130)
	at net.sf.saxon.expr.ValueComparison$ValueComparisonElaborator.lambda$elaborateForItem$1(ValueComparison.java:546)
	at net.sf.saxon.expr.elab.ItemElaborator.lambda$elaborateForPull$0(ItemElaborator.java:36)
	at net.sf.saxon.expr.FilterExpression$FilterExprElaborator$PositionalFilteredIterator.next(FilterExpression.java:1331)
	at net.sourceforge.pmd.lang.rule.xpath.internal.SaxonXPathRuleQuery.evaluate(SaxonXPathRuleQuery.java:127)
	at net.sourceforge.pmd.lang.rule.xpath.XPathRule.apply(XPathRule.java:101)
	at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.applyOnIndex(RuleApplicator.java:77)
	... 9 more
Exception occurred on node !debug only! [MethodCall:113:17]logger.atInfo().log("Got no {} restrictions injected")
```

**Code Sample demonstrating the issue:**

```java
logger.atInfo().log("Got no {} restrictions injected");
```

**Steps to reproduce:**

```xpath
//MethodCall
[@MethodName = "log"]
[
    *[1]/descendant::VariableAccess[pmd-java:typeIs("org.apache.logging.log4j.Logger")]
    or *[1]/descendant-or-self::VariableAccess[pmd-java:typeIs("org.apache.logging.log4j.LogBuilder")]
]
[count(tokenize(ArgumentList/*[1]/self::StringLiteral/@LiteralText, "\{\}")) != count(ArgumentList/*)]
```

**Running PMD through:** *Gradle*

--- a/pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/xpath/Attribute.java
+++ b/pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/xpath/Attribute.java
@@ -8,12 +8,14 @@ import java.lang.invoke.MethodHandle;
 import java.lang.reflect.Method;
 import java.lang.reflect.Type;
 import java.util.Objects;
+import java.util.function.Function;
 
 import org.checkerframework.checker.nullness.qual.NonNull;
 import org.checkerframework.checker.nullness.qual.Nullable;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import net.sourceforge.pmd.annotation.InternalApi;
 import net.sourceforge.pmd.lang.ast.Node;
 
 /**
@@ -47,6 +49,7 @@ public final class Attribute {
 
     /** Must be non-null after {@link #getStringValue()} has been invoked. */
     private String stringValue;
+    private final Function<Method, Type> getType;
 
     /**
      * Creates a new attribute belonging to the given node using its accessor.
@@ -62,16 +65,32 @@ public final class Attribute {
         this.name = Objects.requireNonNull(name);
         this.handle = Objects.requireNonNull(handle);
         this.method = Objects.requireNonNull(method);
+        this.getType = Method::getGenericReturnType;
     }
 
     /** Creates a new attribute belonging to the given node using its string value. */
     public Attribute(@NonNull Node parent, @NonNull String name, @Nullable String value) {
+        this(parent, name, value, m -> String.class);
+    }
+
+    /**
+     * Creates a new attribute belonging to the given node using an arbitrary object value and a type.
+     * Note that the type must be supported by {@link net.sourceforge.pmd.lang.rule.xpath.internal.DomainConversion}
+     * if you want this attribute to be used from XPath.
+     */
+    @InternalApi // used for tests
+    public Attribute(@NonNull Node parent, @NonNull String name, @Nullable Object value, Type type) {
+        this(parent, name, value, m -> type);
+    }
+
+    private Attribute(@NonNull Node parent, @NonNull String name, @Nullable Object value, Function<Method, Type> getType) {
         this.parent = Objects.requireNonNull(parent);
         this.name = Objects.requireNonNull(name);
+        this.getType = getType;
         this.value = value;
         this.handle = null;
         this.method = null;
-        this.stringValue = value == null ? "" : value;
+        this.stringValue = value == null ? "" : value.toString();
         this.invoked = true;
     }
 
@@ -80,7 +99,7 @@ public final class Attribute {
      * Gets the generic type of the value of this attribute.
      */
     public Type getType() {
-        return method == null ? String.class : method.getGenericReturnType();
+        return getType.apply(this.method);
     }
 
     /** Return the name of the attribute (without leading @ sign). */

--- a/pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/xpath/internal/DomainConversion.java
+++ b/pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/xpath/internal/DomainConversion.java
@@ -14,6 +14,7 @@ import org.checkerframework.checker.nullness.qual.NonNull;
 import net.sf.saxon.om.AtomicArray;
 import net.sf.saxon.om.AtomicSequence;
 import net.sf.saxon.om.EmptyAtomicSequence;
+import net.sf.saxon.str.StringTool;
 import net.sf.saxon.type.BuiltInAtomicType;
 import net.sf.saxon.type.SchemaType;
 import net.sf.saxon.value.AtomicValue;
@@ -146,6 +147,8 @@ public final class DomainConversion {
             return new FloatValue((Float) value);
         } else if (value instanceof Pattern || value instanceof Enum) {
             return new StringValue(String.valueOf(value));
+        } else if (value instanceof CharSequence) {
+            return new StringValue(StringTool.fromCharSequence((CharSequence) value));
         } else {
             // We could maybe use UntypedAtomicValue
             throw new RuntimeException("Unable to create ValueRepresentation for value of type: " + value.getClass());

--- a/pmd-core/src/test/java/net/sourceforge/pmd/lang/ast/DummyNode.java
+++ b/pmd-core/src/test/java/net/sourceforge/pmd/lang/ast/DummyNode.java
@@ -4,6 +4,7 @@
 
 package net.sourceforge.pmd.lang.ast;
 
+import java.lang.reflect.Type;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Iterator;
@@ -116,6 +117,9 @@ public class DummyNode extends AbstractNode<DummyNode, DummyNode> {
         attributes.add(new Attribute(this, name, value));
     }
 
+    public void setXPathAttribute(String name, Object value, Type type) {
+        attributes.add(new Attribute(this, name, value, type));
+    }
 
     public void clearXPathAttributes() {
         attributes.clear();

--- a/pmd-core/src/test/java/net/sourceforge/pmd/lang/rule/xpath/internal/SaxonXPathRuleQueryTest.java
+++ b/pmd-core/src/test/java/net/sourceforge/pmd/lang/rule/xpath/internal/SaxonXPathRuleQueryTest.java
@@ -28,6 +28,7 @@ import net.sourceforge.pmd.lang.ast.DummyNode.DummyRootNode;
 import net.sourceforge.pmd.lang.ast.DummyNodeWithListAndEnum;
 import net.sourceforge.pmd.lang.ast.Node;
 import net.sourceforge.pmd.lang.ast.RootNode;
+import net.sourceforge.pmd.lang.document.Chars;
 import net.sourceforge.pmd.lang.rule.xpath.PmdXPathException;
 import net.sourceforge.pmd.lang.rule.xpath.XPathVersion;
 import net.sourceforge.pmd.lang.rule.xpath.impl.XPathHandler;
@@ -138,6 +139,19 @@ class SaxonXPathRuleQueryTest {
                 "//dummyNode[count(distinct-values(@Lines)) > 0 and not(empty(index-of(@Lines, 'a')))]", dummy);
 
         assertEquals(dummy.getChild(0), result.get(0));
+    }
+
+    @Test
+    void testCharsAttributes() {
+        DummyRootNode dummy = helper.parse("(a(b))");
+        Chars value = Chars.wrap("__abcd__").slice(2, 4);
+        //                          ----
+        dummy.setXPathAttribute("CharsAttr", value, Chars.class);
+
+        List<Node> result = assertQuery(1,
+            "//dummyRootNode[@CharsAttr = 'abcd']", dummy);
+
+        assertEquals(dummy, result.get(0));
     }
 
     @Test

